dsgrid/project.py

```diff
@@ -220,9 +201,151 @@ def _build_filtered_record_ids_by_dimension_type(self, context: QueryContext):
+        cached_dataset_path = hash_dir / (dataset.dataset_id + ".parquet")
+        metadata_file = cached_dataset_path.with_suffix(".json5")
+        if try_read_dataframe(cached_dataset_path) is None:
+            self._build_exponential_growth_dataset(
+                context,
+                cached_datasets_dir,
+                dataset,
+                cached_dataset_path,
+                metadata_file,
+            )
+        else:
+            assert metadata_file.exists(), metadata_file
+            context.set_dataset_metadata_from_file(dataset.dataset_id, metadata_file)
+            logger.info("Use cached project-mapped dataset %s", dataset.dataset_id)
+
+        return cached_dataset_path
+
+    @track_timing(timer_stats_collector)
+    def _build_exponential_growth_dataset(
+        self, context, cached_datasets_dir, dataset, dataset_path, metadata_file
+    ):
+        def get_myear_column(dataset_id):
+            names = list(
+                context.get_dimension_query_names(DimensionType.MODEL_YEAR, dataset_id=dataset_id)
+            )
+            assert len(names) == 1, f"{dataset_id=} {names=}"
+            return names[0]
+
+        iv_path = self._process_dataset(
+            context, cached_datasets_dir, dataset.initial_value_dataset_id
+        )
+        gr_path = self._process_dataset(
+            context, cached_datasets_dir, dataset.growth_rate_dataset_id
+        )
+        pivoted_columns = context.get_pivoted_columns(dataset_id=dataset.initial_value_dataset_id)
+        pivoted_columns_gr = context.get_pivoted_columns(dataset_id=dataset.growth_rate_dataset_id)
+        if pivoted_columns != pivoted_columns_gr:
+            raise Exception(
+                f"BUG: Mismatch in pivoted columns: "
+                f"{pivoted_columns.symmetric_difference(pivoted_columns_gr)}"
+            )
+
+        model_year_column = get_myear_column(dataset.initial_value_dataset_id)
+        model_year_column_gr = get_myear_column(dataset.growth_rate_dataset_id)
+        if model_year_column != model_year_column_gr:
+            raise Exception(
+                "BUG: initial_value and growth rate datasets have different model_year columns: "
+                f"{model_year_column=} {model_year_column_gr=}"
+            )
+        time_columns = context.get_dimension_query_names(
+            DimensionType.TIME, dataset_id=dataset.initial_value_dataset_id
+        )
+        with restart_spark_with_custom_conf(
+            conf=context.model.project.get_spark_conf(dataset.dataset_id),
+            force=True,
+        ):
+            logger.info("Build projection dataset %s", dataset.dataset_id)
+            iv_df = read_dataframe(iv_path)
+            gr_df = read_dataframe(gr_path)
+            if dataset.construction_method == "formula123":
+                df = apply_growth_rate_123(
+                    dataset, iv_df, gr_df, time_columns, model_year_column, pivoted_columns
+                )
```

Comment:
Ick. Seems like the string tag should match the function name? "apply_growth_rate_123" not "formula123"? And there should at least be a dict somewhere there that provides the look up?

Reply:
This is a placeholder. @lixiliu agreed to create a naming convention in a subsequent PR.

Reply:
okay

Reply:
Perhaps I should not have deleted dsgrid_project/datasets/modeled/resstock/dimension_mappings/model_year_to_model_year.csv in https://github.com/dsgrid/dsgrid-project-StandardScenarios/pull/71.